### PR TITLE
cluster builder interface and updated output

### PR DESCRIPTION
**Pull request by dhavalpatel624624**
## Project Robyn
Generated cluster builder interface which will cluster models and updated the clustering_result dataclass for outputs

# No testing needed at this point

### CODE REVIEW

File: python/src/robyn/modeling/clustering/cluster_builder.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        Args:
+            clustered_df (pd.DataFrame): The DataFrame containing cluster assignments.
+            config (ClusteringConfig): Configuration for the clustering process.
+
+        Returns:
+            pd.DataFrame: DataFrame of top models.
+        """
+        pass
+
+    def plot_confidence_intervals(
```
Comment (alxlyj) [inline python/src/robyn/modeling/clustering/cluster_builder.py:129]: let's take out plot from cluster component here. I see some of the methods here already located at cluster_visualization.py https://github.com/facebookexperimental/Robyn/blob/robynpy_release/python/src/robyn/visualization/cluster_visualization.py

File: python/src/robyn/modeling/clustering/cluster_builder.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        Args:
+            confidence_data (Dict[str, float]): The data containing confidence intervals for plotting.
+            config (ClusteringConfig): Configuration for the clustering process.
+
+        Returns:
+            None
+        """
+        pass
+
+    def plot_top_solutions(self, config: ClusteringConfig) -> None:
```
Comment (alxlyj) [inline python/src/robyn/modeling/clustering/cluster_builder.py:144]: ditto

File: python/src/robyn/modeling/entities/clustering_results.py [hunk 3]
```diff
@@ -1,64 +1,87 @@
...
+import matplotlib.pyplot as plt
+
 import pandas as pd
-import numpy as np
-from typing import List
 
 
 @dataclass
-class KMeansResults:
+class PlotResults:
```
Comment (alxlyj) [inline python/src/robyn/modeling/entities/clustering_results.py:11]: if these are all generated during clustering, perhaps we can consider renaming this to ClusteringPlotResults?

File: python/src/robyn/modeling/entities/clustering_results.py [hunk 4]
```diff
@@ -1,64 +1,87 @@
...
-        centroids (np.ndarray): Coordinates of cluster centroids.
-        inertia (float): Sum of squared distances of samples to their closest cluster center.
+        data (pd.DataFrame): The DataFrame containing the clustered models.
+        top_solutions (pd.DataFrame): The top solutions based on clustering.
     """
 
-    cluster_assignments: List[int]
-    centroids: np.ndarray
-    inertia: float
+    data: pd.DataFrame
```
Comment (alxlyj) [inline python/src/robyn/modeling/entities/clustering_results.py:42]: do you think we could avoid generics like data?

File: python/src/robyn/modeling/entities/clustering_results.py [hunk 5]
```diff
@@ -1,64 +1,87 @@
...
-
-    def export_results(self, path: str) -> None:
-        """
-        Export the clustering results to files.
-
-        Args:
-            path (str): Directory path to save the results.
-        """
-        pass
+    cluster_data: ClusterData
```
Comment (alxlyj) [inline python/src/robyn/modeling/entities/clustering_results.py:79]: is there a distinction here between ClusterData.data and ClusteringResults.data? could we be more specific here?

**Review by alxlyj — COMMENTED** [review]
few comments primarily on namings and refactoring for visualization. we can look into getting rid of the older clustering interfaces in our next iteration.

**Comment by sumane81** [comment]
I asked AI to review the code for me.
_______________________________
I'll review cluster_builder.py and clustering_results.py focusing on code structure, encapsulation, type safety and proper initialization. Let me analyze each file and suggest improvements.

First, let's examine clustering_results.py:

Key improvements to clustering_results.py:
1. Added `frozen=True` to make dataclasses immutable
2. Added proper type annotations for all attributes
3. Used `field(default_factory=...)` for mutable defaults 
4. Added comprehensive docstrings
5. Made all dataclasses properly initialized with defaults

Now let's examine cluster_builder.py. This class has multiple responsibilities and should be split up:

Key improvements to cluster_builder.py:

1. Split into specialized classes following Single Responsibility Principle:
   - DataPreprocessor: Handles data preparation
   - ClusterAnalyzer: Performs statistical analysis
   - ConfidenceIntervalCalculator: Calculates confidence intervals
   - ClusterBuilder: Coordinates the overall process

2. Improved encapsulation:
   - Made internal methods private with underscore prefix
   - Added proper type hints
   - Used dataclasses for data structures
   - Protected internal state

3. Better organization:
   - Each class has a single, clear responsibility
   - Methods are focused and cohesive
   - Clear separation of concerns

4. Additional improvements:
   - Added comprehensive docstrings
   - Made method signatures explicit
   - Used proper type annotations throughout
   - Made the code more maintainable and testable

The refactored code is now:
1. More maintainable due to clear separation of concerns
2. More robust with proper type checking
3. Better encapsulated with clear public/private interfaces
4. Easier to test due to smaller, focused classes
5. More flexible for future modifications
[cluster-builder-refactor.MD](https://github.com/user-attachments/files/17515770/cluster-builder-refactor.MD)
[clustering-results-refactor.MD](https://github.com/user-attachments/files/17515771/clustering-results-refactor.MD)

___________________________


Attached refactored code. Change .MD extension to .py

**Review by sumane81 — APPROVED** [review]
Looks good to me.

**Pull request merged**